### ISSUE TITLE
another shaper issues in arabic

### ISSUE THREAD

**Issue by realfinder** [libass/libass#71]
hi

using last mpv in windows (with HarfBuzz)

libass
![1 libass](https://cloud.githubusercontent.com/assets/2512153/2575335/f1be0b5c-b94e-11e3-8ae0-540b618aea4d.png)

libass with the Pointing
![1 libass a](https://cloud.githubusercontent.com/assets/2512153/2575386/2d3fb9ae-b950-11e3-99ec-6abb588bb13c.png)

vsfilter
![1 vsfilter](https://cloud.githubusercontent.com/assets/2512153/2575337/f1cb0208-b94e-11e3-87c9-6dc391b26953.png)

---

libass
![2 libass](https://cloud.githubusercontent.com/assets/2512153/2575336/f1c60ae6-b94e-11e3-813c-8ddb3de84d62.png)

vsfilter
![2 vsfilter](https://cloud.githubusercontent.com/assets/2512153/2575334/f1bc7d6e-b94e-11e3-8ae8-12ed93dc7dbb.png)

wrong word order because of underline code

---

video sample http://www.solidfiles.com/d/9287427ae4/[Devil-Fansub]_Hyouka_-_11.mkv

sub + font http://www.solidfiles.com/d/1c8cb1aace/sub_%2B_font.rar

thanks

**Comment by grigorig** [comment]
Regarding the "wrong order", at the moment, the behaviour on override tags is somewhat intentional incompatibility with VSFilter. As far as I can tell VSFilter's behaviour (always reset to LTR after an override) is really unusual and inconsistent compared to BiDi interaction in other markup languages like HTML. Do we really want to change that?

I'll take a look at the other problems. We should especially filter out control characters much earlier. (Font fallback happens way before shaping)

**Comment by realfinder** [comment]
hi again

since a lot of people used mobile now, and I saw there are some complaints about these compatibility cases (like in this "another shaper issues in arabic #71" and others compatibility issues)

any intention to make libass more compatible with vsfilter?

**Comment by grigorig** [comment]
I'm not sure, VSFilter behavior is very inconsistent. We've already done various changes to make libass behave closer to VSFilter, but there's a limit. You proposed to reset the text direction at each override tag, which seems like a pretty significant change. Let me think about it.

**Comment by astiob** [comment]
Friendly reminder: what VSFilter does is split the text into character runs, lay out the character runs from left to right, and then render each run separately. libass currently performs the same splitting, but only after almost all rendering is done (producing a single “combined bitmap” for each run). In particular, this happens after glyphs are laid out (applying the bidirectional algorithm to the entire text at once). In fact, this discrepancy causes additional compatibility issues besides bidirectional layout, so changing libass to split runs much earlier would bring additional benefits, but it is not a trivial task.

**Comment by wm4** [comment]
We could also try to push sane semantics (and hopefully get Aegisub's help?), and let VSFilter rot in the corner.

**Comment by kirito90** [comment]
> since a lot of people used mobile now, and I saw there are some complaints about these compatibility cases (like in this "another shaper issues in arabic #71" and others compatibility issues)

same happen in Kurdish (Sorani) too

**Comment by astiob** [comment]
@realfinder Do Arabic/Hebrew subtitles come in SRT format with bold/underline/italics as well? Are they popular? Do they use VSFilter’s nonsensical word order, too?

**Comment by realfinder** [comment]
not sure, from what I see in subscene.com non anime movies and series translators used SRT with even Windows-1256 code page, and not all of them use same methods, but I think yes, they made it with VSFilter in mind
![Untitled](https://user-images.githubusercontent.com/2512153/77391836-c5e1d100-6daa-11ea-92f5-ace0d5125b93.png)
Joker.2019.1080p.BrRip.2CH.x265.HEVC-PSA {Not-Colored}.srt one is ANSI

**Comment by realfinder** [comment]
this fixed with https://github.com/libass/libass/commit/c629be7d70548aea282ea4890c51094055ec66e2 also I note that https://github.com/libass/libass/issues/71#issuecomment-48186912 is also ok in last mpv in windows

other issues are here https://github.com/harfbuzz/harfbuzz/issues/2888 now
